### PR TITLE
Refactor handleCallInst function.

### PR DESCRIPTION
**Pull request by ptersilie**
More commits simplifying and improving readability of the `jitmodbuilder` code, in particular `handleCallInst`.

### CODE REVIEW

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 1]
```diff
@@ -95,6 +95,17 @@ class JITModBuilder {
     }
   }
 
+  bool isRecursiveCall(Function *F) {
```
Comment (ltratt) [inline ykllvmwrap/src/jitmodbuilder.cc:94]: Can we add a doc string for this?

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 2]
```diff
@@ -123,31 +134,26 @@ class JITModBuilder {
...
+        if (VMap[CF] == nullptr) {
+          StringRef CFName = CF->getName();
+          // Declare function.
+          auto DeclFunc = llvm::Function::Create(CF->getFunctionType(),
+                                                 GlobalValue::ExternalLinkage,
+                                                 CFName, JITMod);
+          VMap[CF] = DeclFunc;
+          for (size_t i = 0; i < FAddrLen; i++) {
+            char *FName = FAddrKeys[i];
+            uint64_t FAddr = FAddrVals[i];
```
Comment (ltratt) [inline ykllvmwrap/src/jitmodbuilder.cc]: Is it worth breaking this out as a separate variable? It only seems to be used once (on line 149)?

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 3]
```diff
@@ -137,11 +141,7 @@ class JITModBuilder {
       if (isRecursiveCall(CF)) {
         if (VMap[CF] == nullptr) {
           StringRef CFName = CF->getName();
-          // Declare function.
-          auto DeclFunc = llvm::Function::Create(CF->getFunctionType(),
-                                                 GlobalValue::ExternalLinkage,
-                                                 CFName, JITMod);
-          VMap[CF] = DeclFunc;
+          declareFunction(CF);
```
Comment (ltratt) [inline ykllvmwrap/src/jitmodbuilder.cc:151]: Possibly silly question: can the same function be declared on both lines 121 and 144?

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 4]
```diff
@@ -113,6 +113,18 @@ class JITModBuilder {
     VMap[F] = DeclFunc;
   }
 
+  void addGlobalMappingForFunction(Function *CF) {
```
Comment (ltratt) [inline ykllvmwrap/src/jitmodbuilder.cc:112]: Let's add a docstring.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 3 above)
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:151]: If you declare a duplicate function name, LLVM will add a de-duplications suffix to the symbol name.

That said, I think `declareFunction()` should check so as to not bloat up the IR. It should probably also `assert` that the signature is the same?

Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:151]: > That said, I think declareFunction() should check so as to not bloat up the IR. It should probably also assert that the signature is the same?

Agreed on both points. [Although -- if it's possible -- better than checking if the function has been declared would be to work out how not to double declare it in the first place?]

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 5]
```diff
@@ -67,11 +67,7 @@ class JITModBuilder {
   // Instruction at which to continue after an a call.
   Optional<tuple<size_t, CallInst *>> ResumeAfter;
   // Depth of nested calls when outlining a recursive function.
-  // FIXME: can we kill this?
   size_t RecCallDepth = 0;
-  // Function currently being outlined.
-  // FIXME: this should be an Optional?
-  tuple<size_t, CallInst *> NoInlineFunc;
```
Comment (vext01) [inline ykllvmwrap/src/jitmodbuilder.cc:74]: Brilliant!

**Comment by vext01** [comment]
Not much to add on top of what laurie has already said.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 3 above)
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:151]: It's inevitable when the trace calls the same symbol a second time.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 2 above)
Comment (ptersilie) [reply ykllvmwrap/src/jitmodbuilder.cc]: I find it improves readability. But I don't feel strongly about it. `FName` is also only used once.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 3 above)
Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:151]: Ah!

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 2 above)
Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc]: I don't feel strongly either, so do as you see fit.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 3 above)
Comment (ptersilie) [reply ykllvmwrap/src/jitmodbuilder.cc:151]: We already guard against that with `if (VMap[CF] == nullptr) {`. We never call `declareFunction` if it's already in the `VMap`.

Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:151]: Even better. Still, an `assert` to make this clear in `declareFunction` might be a good idea.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 4 above)
Comment (ptersilie) [reply ykllvmwrap/src/jitmodbuilder.cc:117]: Fixed in 27db37f.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 1 above)
Comment (ptersilie) [reply ykllvmwrap/src/jitmodbuilder.cc:96]: Fixed in f71b25b.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 3 above)
Comment (ptersilie) [reply ykllvmwrap/src/jitmodbuilder.cc:156]: I think d730754 should do the trick. See if you agree at @vext01.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 2 above)
Comment (ptersilie) [reply ykllvmwrap/src/jitmodbuilder.cc]: I think I prefer it as is.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 3 above)
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:156]: LGTM

**Comment by ptersilie** [comment]
> Based on another discussion, I think this function's docstring might better be something like "Find the machine code corresponding to the AOT IR `CF` and ensure there's an entry from `CF`'s name to that machine code".

Try ed59f3e.

**Comment by ltratt** [comment]
Please squash.

**Comment by ptersilie** [comment]
Squashed.

**Comment by ptersilie** [comment]
Forgot to push a formatting change.

**Comment by ptersilie** [comment]
There it is.

**Comment by ltratt** [comment]
Please squash.

**Comment by ptersilie** [comment]
Squashed.

**Pull request merged**